tests/network/sriov.go

```diff
@@ -695,3 +657,157 @@ func findIfaceByMAC(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineIns
 	}
 	return ifaceName, nil
 }
+
+func newSRIOVVmi(networks []string, cloudInitNetworkData string) *v1.VirtualMachineInstance {
+	options := []libvmi.Option{
+		libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+		libvmi.WithNetwork(v1.DefaultPodNetwork()),
+	}
+	if cloudInitNetworkData != "" {
+		cloudinitOption := libvmi.WithCloudInitNoCloudNetworkData(cloudInitNetworkData)
+		options = append(options, cloudinitOption)
+	}
+
+	for _, name := range networks {
+		options = append(options,
+			libvmi.WithInterface(libvmi.InterfaceDeviceWithSRIOVBinding(name)),
+			libvmi.WithNetwork(libvmi.MultusNetwork(name, name)),
+		)
+	}
+	return libvmi.NewFedora(options...)
+}
+
+func checkInterfacesInGuest(vmi *v1.VirtualMachineInstance, interfaces []string) {
+	for _, iface := range interfaces {
+		Expect(checkInterface(vmi, iface)).To(Succeed())
+	}
+}
+
+// createAndWaitVMI creates the received VMI and waits for the guest to load the guest-agent
+func createAndWaitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
+	virtClient, err := kubecli.GetKubevirtClient()
+	if err != nil {
+		panic(err)
+	}
+
+	vmi, err = virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(vmi)
+	if err != nil {
+		return nil, err
+	}
+
+	return waitVMI(vmi)
+}
+
+func waitVMI(vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
+	// Running multi sriov jobs with Kind, DinD is resource extensive, causing DeadlineExceeded transient warning
+	// Kubevirt re-enqueue the request once it happens, so its safe to ignore this warning.
+	// see https://github.com/kubevirt/kubevirt/issues/5027
+	warningsIgnoreList := []string{"unknown error encountered sending command SyncVMI: rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
+	tests.WaitUntilVMIReadyIgnoreSelectedWarnings(vmi, console.LoginToFedora, warningsIgnoreList)
+
+	virtClient, err := kubecli.GetKubevirtClient()
+	if err != nil {
+		panic(err)
+	}
+
+	tests.WaitAgentConnected(virtClient, vmi)
+
+	return virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &k8smetav1.GetOptions{})
+}
+
+// deleteVMI deletes the specified VMI and waits for its absence.
```

Comment:
My understanding is this is a bug so please let's ensure we have an issue, todo... to remove this hack in the future.

Reply:
I consider the global cleanup to be wrong, so I consider this to be the right way to write tests.
But if you insist, I will add it.

Reply:
I am with Edy here, we never block upon delete, its not PR specific

Reply:
That is another topic and we should solve it in a general way imho. WDYT?

Reply:
The question remains:
> Should we wait for all VMI/s to be gone before declaring the test as finished?

I argue we should, in all cases, making sure we do not start a test while older objects exist in the cluster, which may have unclear expectations.

But for the SR-IOV tests specifically, I think the requirement is even stronger.
The resources each VMI uses are HW related and keeping old ones around may causes issues, like the one we see now.
It is true that it may occur due to a bug, but we should asks ourself if this stress test was indeed explicitly intended or just a side effect. I think we can agree it is a side-effect.

The only general rule I can think of, is this: If a test creates objects that use cluster or node wide resources, it will be correct to wait for the object removal before declaring the test over.
As SR-IOV VMI/s answer this at the node resource level, I think the suggested change here can be consider a good solution, both to the current problem we sometime face and to the general proposed "rule".

But again, if you are not convinced, and this is will be reason to not push this PR forward, I will add the ref to a the bug.

Reply:
All of these are imo requirements for the testing framework @fossedihelm proposed - some stuff can be lazily deleted (assuming we'd be using a different namespace for each test) but some - like sriov resources - **must** be synchronously deleted on the test teardown.

Are these scenarios considered on your proposal @fossedihelm ? 

... I'm aware I owe you a review on that design doc ...

Reply:
@maiqueb I didn't consider these scenarios, so we should expose a way to set a specific behavior: Wait or don't wait resources deletion. Thanks for pointing me this!

Reply:
> My understanding is this is a bug so please let's ensure we have an issue, todo... to remove this hack in the future.

I agree with this comment. There is a real bug and we are covering it with this cleanup. Please add a comment with a reference to the bug opened on the relevant component (SRIOV operator or whatever component). If it is still under investigation what is the corresponding component, please open a kubevirt issue.
Even though it is a corner case, it is still a real bug.

Reply:
We suspect there is a bug caused by this behavior. But it does not matter, our tests are not here to test this scenario (load of reallocations) and cleaning up properly as suggested here is the correct thing, even without this potential bug.

adding such a ref implies we can leave vmi/s around in “deleting” state, which is IMO wrong.

Reply:
Yes, our tests are designed this way. A VM from a previous test can be in deleting state when we run a new test.
In most of the tests we are not waiting for the previous VMs to be deleted.
Here we need to wait due to a bug and not because it is the correct test design.
Therefore, we should document it is a workaround.
And if you don't consider it as a workaround then the official documentation should be updated to tell the user has to wait for a previous VM using a resouce to be removed before running a new VM using/potentially using the same resource.

Reply:
Let me just add a few points:
1. This is a difference in how we do clean-up in the test suite. It is irelevant if this is the "right" way to do it or not.
2. We would not be making this change if we would not hit this bug. Hiding this fact doesn't do right to anyone.

Reply:
I would personally be for having it here. @AlonaKaplan I let you decide.

Reply:
Tracking issue https://github.com/k8snetworkplumbingwg/sriov-cni/issues/219.

As a side note, I agree that the way VMIs are being cleaned up should be revisited.
The fact that we dispose VMIs by removing their finalizers is in general bad practice and could lead to resource leaks.
https://github.com/kubevirt/kubevirt/blob/28630a3f91434203f2eac3f4c5bce217e6a47833/tests/testsuite/namespace.go#L90-L101

Reply:
I believe that the correctness of our current cleanup code is irrelevant here.
In this case it helped us find a real bug, so maybe it is not that bad?

The tests here were changed to have a non-standard cleanup due to a bug.
Please add the reference to the bug to the tests code (and not just the commit message) so it will be easier for the next developer that touches it to decide whether the workaround is still needed or not.

Reply:
I added the reference.
Hope this is enough to close this point.